extensions/amp-analytics/0.1/scroll-manager.js

```diff
@@ -99,14 +129,19 @@ export class ScrollManager {
    * @private
    */
   onScroll_(e) {
+    // Handle offsets for shadow docs
+    const layoutRect = this.viewport_.getLayoutRect(this.root_);
+
     /** {./scroll-manager.ScrollEventDef} */
     const scrollEvent = {
-      top: e.top,
-      left: e.left,
+      top: e.top - layoutRect.top,
```

Comment:
same question, could you please expand a bit on what are the `this.viewport_.getScrollTop()`, `layoutRect.top`. thanks

Reply:
This shouldn't be a breaking change because for the top-level page, `this.root_` is equivalent to the document body which means that for the existing case, `layoutRect.top` and `layoutRect.left` would be `0` (so it doesn't change the existing behavior). For the ShadowDOM case, `this.root_` becomes the ShadowDOM's container, in which case the `layoutRect.top` and `layoutRect.left` would serve to offset the calculations to become relative to the ShadowDOM's container as opposed to the top level container.

Reply:
@zhouyx are there any test suites / manual tests we can do here to ensure that nothing breaks?

Reply:
nothing that tests shadow dom that I'm aware of. Please feel free to create and add new ones. Thanks!

Reply:
Thanks for the explanation. One more question on the `viewport.getScrollTop()` for the non-shadow case. Does the shadow doc share the same viewport service as the parent? What is the scrollTop value here?

Reply:
If I understand correctly, it does not, the shadow docs will have a viewport service for their own ampdoc instance ( `AmpDocShadow` as opposed to `AmpDocSingle`) so `viewport.getScrollTop` will yield the distance between top of the page and the position of the shadow doc.